lightning/src/ln/features.rs

```diff
@@ -18,17 +18,106 @@ use std::marker::PhantomData;
 use ln::msgs::DecodeError;
 use util::ser::{Readable, Writeable, Writer};
 
-#[macro_use]
-mod sealed { // You should just use the type aliases instead.
-	pub struct InitContext {}
-	pub struct NodeContext {}
-	pub struct ChannelContext {}
-
-	/// An internal trait capturing the various feature context types
-	pub trait Context {}
-	impl Context for InitContext {}
-	impl Context for NodeContext {}
-	impl Context for ChannelContext {}
+mod sealed {
+	/// The context in which [`Features`] are applicable. Defines which features are supported.
+	///
+	/// [`Features`]: ../struct.Features.html
+	pub trait Context {
+		/// Features that are supported by the implementation, indicated by setting their
+		/// optional (odd) bits.
+		const SUPPORTED_FEATURE_FLAGS: &'static [u8];
+
+		/// Bitmask for selecting features that are known though not necessarily implemented.
```

Comment:
Hmm, this may be a good opportunity to update the naming - 'known' is maybe somewhat misleading, its more like 'supported-by-the-remote-party-setting-the-bit'. eg we don't add all the features specified if we don't know how to talk to a peer that requires it.

Reply:
This is only used by `with_known_relevant_init_flags` which takes `InitFeatures` from a counterparty (i.e. 'supported-by-the-remote-party-setting-the-bit') and returns features we know about (i.e. 'we don't add all the features specified if we don't know how to talk to a peer that requires it').

So I think this describes the intention accurately based on what you said and [comment](https://github.com/rust-bitcoin/rust-lightning/blob/master/lightning/src/ln/features.rs#L159-L162) within `with_known_relevant_init_flags`. Do I have that right?

Come to think of it, I can probably drop `KNOWN_FEATURE_MASK` entirely and bitwise negate each byte in `UNKNOWN_FEATURE_MASK` when used within `with_known_relevant_init_flags` instead. Let me know if you'd prefer that.

Reply:
I mostly meant that 'known' implies also including bits which we know about, but don't even support it (eg channel queries). Known is a strange name here. `with_known_relevant_init_flags` is a bit more clear in that its "known" and "relevant in both init and the context we're copying to". Dunno what the answer is, really.

Reply:
From our offline discussion, we will treat "known" and "supported" as synonymous and only use the term "known".

Instead of defining contexts in terms of `supported_features` and `unsupported_features`, I've defined them using `required_features` and `optional_features`. This allows us to make features required in the future.